File: src/main/java/mediathek/x11/DesktopEnvDetector.java
package mediathek.x11;

import org.apache.commons.lang3.SystemUtils;

import java.io.IOException;
import java.util.Locale;
import java.util.Map;

public final class DesktopEnvDetector {

    public enum DesktopEnvironment {
        KDE, GNOME, UNITY, UNKNOWN
    }

    private DesktopEnvDetector() {
    }

    public static boolean trayIconSupported() {
        if (!SystemUtils.IS_OS_LINUX)
            return true;

        return (detect() != DesktopEnvironment.KDE);
    }

    public static DesktopEnvironment detect() {
        String envValue = getDesktopEnvFromVariables().toUpperCase(Locale.ROOT);

        if (envValue.contains("KDE"))
            return DesktopEnvironment.KDE;
        else if (envValue.contains("GNOME"))
            return DesktopEnvironment.GNOME;
        else if (envValue.contains("UNITY"))
            return DesktopEnvironment.UNITY;

        // Fallback
        if (isProcessRunning("plasmashell"))
            return DesktopEnvironment.KDE;
        else if (isProcessRunning("gnome-shell"))
            return DesktopEnvironment.GNOME;
        else if (isProcessRunning("unity-panel-service"))
            return DesktopEnvironment.UNITY;

        return DesktopEnvironment.UNKNOWN;
    }

    private static String getDesktopEnvFromVariables() {
        Map<String, String> env = System.getenv();

        if (env.containsKey("XDG_CURRENT_DESKTOP"))
            return env.get("XDG_CURRENT_DESKTOP");
        else if (env.containsKey("DESKTOP_SESSION"))
            return env.get("DESKTOP_SESSION");
        else if (env.containsKey("GDMSESSION"))
            return env.get("GDMSESSION");
        else if (env.containsKey("GNOME_DESKTOP_SESSION_ID"))
            return "GNOME";
        else
            return "unknown";
    }

    private static boolean isProcessRunning(String process) {
        try {
            Process p = new ProcessBuilder("pgrep", "-x", process)
                    .redirectErrorStream(true)
                    .start();
            int exitCode = p.waitFor();
            return exitCode == 0;
        } catch (IOException | InterruptedException e) {
            return false;
        }
    }
}

File: src/main/kotlin/mediathek/x11/DesktopEnvDetector.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.x11

import kotlinx.coroutines.*
import org.apache.commons.lang3.SystemUtils
import java.io.IOException
import java.util.*

object DesktopEnvDetector {
    enum class DesktopEnvironment {
        KDE,
        GNOME,
        UNITY,
        UNKNOWN,
    }

    @JvmStatic
    fun trayIconSupported(): Boolean {
        if (!SystemUtils.IS_OS_LINUX) {
            return true
        }

        return detect() != DesktopEnvironment.KDE
    }

    @JvmStatic
    fun detect(): DesktopEnvironment = runBlocking {
        detectDesktopEnvironment()
    }

    @JvmSynthetic
    suspend fun detectDesktopEnvironment(): DesktopEnvironment {
        val envValue = getDesktopEnvFromVariables().uppercase(Locale.ROOT)

        return when {
            "KDE" in envValue -> DesktopEnvironment.KDE
            "GNOME" in envValue -> DesktopEnvironment.GNOME
            "UNITY" in envValue -> DesktopEnvironment.UNITY
            else -> detectDesktopEnvironmentFromProcesses()
        }
    }

    private fun getDesktopEnvFromVariables(): String {
        val env = System.getenv()

        return env["XDG_CURRENT_DESKTOP"]
            ?: env["DESKTOP_SESSION"]
            ?: env["GDMSESSION"]
            ?: env["GNOME_DESKTOP_SESSION_ID"]?.let { "GNOME" }
            ?: "unknown"
    }

    private suspend fun detectDesktopEnvironmentFromProcesses(): DesktopEnvironment = coroutineScope {
        val processChecks = listOf(
            DesktopEnvironment.KDE to async { isProcessRunning("plasmashell") },
            DesktopEnvironment.GNOME to async { isProcessRunning("gnome-shell") },
            DesktopEnvironment.UNITY to async { isProcessRunning("unity-panel-service") },
        )

        processChecks.firstOrNull { (_, isRunning) -> isRunning.await() }?.first
            ?: DesktopEnvironment.UNKNOWN
    }

    private suspend fun isProcessRunning(process: String): Boolean = withContext(Dispatchers.IO) {
        try {
            val p = ProcessBuilder("pgrep", "-x", process)
                .redirectErrorStream(true)
                .start()
            p.waitFor() == 0
        } catch (_: IOException) {
            false
        } catch (_: InterruptedException) {
            false
        }
    }
}
